src/v/archival/ntp_archiver_service.cc

```diff
@@ -1020,7 +1155,22 @@ ss::future<ntp_archiver::upload_group_result> ntp_archiver::wait_uploads(
           segment_kind);
         co_return total;
     }
-    auto results = co_await ss::when_all_succeed(begin(flist), end(flist));
+
+    auto stm_was_clean = _parent.archival_meta_stm()->get_dirty(
+                           _projected_manifest_clean_at)
+                         == cluster::archival_metadata_stm::state_dirty::clean;
+
+    auto [segment_results, manifest_clean_offset]
```

Comment:
nit: this looks a bit confusing (nested `when_all_succeed`), alternatively, the inline manifest upload can be added to the `flist`

```
if (inline_manifest && _manifest_upload_interval().has_value()) {
  flist.emplace_back(maybe_upload_manifest().then([] (std::optional<model::offset>) { ... convert to upload result and copy clean offset ... }));
}
```

Reply:
I wasn't going to do it because I didn't like carrying a coroutine stack var reference into the lambda.  However, I later added the _projected_manifest_clean_at on the class, which means that the return value of maybe_upload_manifest isn't needed anyway: when we later decide whether to write it to the stm, we can just check the projected clean.

So now the suggestions works great: I've added a commit that simplifies it to include the future in `flist` and avoid returning the offset from `maybe_upload_manifest` at all.